### ISSUE TITLE
API reformulation

### ISSUE THREAD

**Issue by MartinPerez** [nistats/nistats#65]
A recent discussion pointed at several current weaknesses of the API:

+ The error prone need to reinput parameters. For example to instantiate design matrices at different pipeline steps.
+ Lack of separation between data and parameters. Like the relationship between TR and design. We might want to separate design_spec from design_matrix.
+ Misnaming of certain functions in the analysis process. Like the transform in the GLM object that can be misleading. Other functions might be added to the GLM object to get certain types of ouputs.
+ Too much historical baggage. We might want to experiment with different conventions. Like replacing model specification with contrast vectors by an R/Statsmodels type of specification.

The importance of getting the second level analysis object right was emphasized. There were some issues raised related to mask application at first or second level and how to handle complex designs (ex. factorial design). We have to balance carefully user surprisal with limiting the power of the framework for mere historical reasons.

Here a prototype proposal of API changes:

paradigm = load_paradigm('paradigm_sub1.csv')
confounds = pandas.read_csv('motion.csv')

model = GLM(hrf_model="spm", drift_model="", t_r=None,
            mask_img=False)
model.fit(paradigm, epi_imgs, confounds)
print(model.summary())
model.compute_contrast('seen - unseen')

second_level = SecondLevel(type="ffx", mask='auto')
second_level.fit([model1, model2, model3], design=None)
second_level.compute_contrast('(seen - unseen) * (schizo - control)')
second_level.compute_contrast(first='(seen - unseen)'
                              second='(schizo - control)')

**Comment by bthirion** [comment]
Namy thx for the summary. Note that I'm working remotely today. I'll 
start by consider a bit mroe carefully patsy/statsmodel.
Best,

B

**Comment by MartinPerez** [comment]
I will check this out too and bring some code proposal for march 7.

**Comment by MartinPerez** [comment]
After discussion, @bthirion and me arrived to the API reformulation proposal posted in the wiki. https://github.com/nistats/nistats/wiki/API-proposal. Please give us your feedback!

Here a simple uncommented example as teaser. Please check the wiki for detailed comments on the proposal.


`design_matrix_glm_kwargs = {'hrf_model': 'canonical', 'drift_model': 'cosine',`
`                            'low_pass': None, 'high_pass': None}`

`model_formula = 'y = cond1 + cond2 + cond3 + ar_noise'`
`firstlevel = FirstLevelModel('task_id', formula=model_formula, mask_img=None, n_jobs=n_proc,`  
`                             **dict(design_matrix_glm_kwargs,))`

`firstlevel.save(firstlevel_path)`

`dataset = load_bids_dataset(bids_path)`

`firstlevel.fit(dataset)`
`firstlevel.fit(paradigms, epi_imgs, confounds)`

`glm_report(firstlevel)`
`print firstlevel.summary()`

`firstlevel.compute_contrast('contrast_1', 'cond1 - cond2')`

Regressors can be ['main', 'time', 'dispersion', 'main and time', 'main and time and dispersion']. 'main' by default

`firstlevel.compute_contrast('contrast_2', '0.5 * cond3 + 0.5 * cond4', regressors='time')`

`firstlevel.compute_contrast('contrast_3', 'contrast_1 - contrast_2')`

`firstlevel.compute_contrast('motion_test', 'np.vstack(m1, m2, m3, m4, m5, m6)')`

`firstlevel.save()`
`contrast_list = firstlevel.get_contrast_definition()`

`firstlevel.compute_contrast(contrasts='cond1', output=['effects', 'variance'])`
`firstlevel.compute_contrast(contrasts='contrast1', output=['effects', 'variance'])`
`firstlevel.compute_contrast(contrasts='all', output=['stat', 'p_value'],`
`                            non_parametric=True)`

`model_formula = 'y = contrast + age + IQ'`
`second_level = SecondLevelModel('task_id', formula=model_formula,`
`                                type='random effects',`
`                                mask_img=None, n_jobs=n_proc,`
`                                **kwargs)`

`second_level.save(second_level_path)`

`second_level.fit(first_level, second_level_covariates)`

`second_level.save(second_level_path)`

`second_level.compute_contrast(contrast='contrast_id * intercept',`
`                              output=['stat', 'p_value'])`
`second_level.compute_contrast(contrast='contrast_id * age',`
`                              output=['stat', 'p_value'])`

`glm_report(secondlevel)`  
`print secondlevel.summary()`

**Comment by TheChymera** [comment]
Is the BIDS organization of the data a hard requirement, or can I somehow pass the paths to my preprocessed images to the function?

**Comment by MartinPerez** [comment]
you could also just pass a list of the images

**Comment by chrplr** [comment]
Hi Martin,

The API must be clear about the difference between factors and the levels of factors. 
In your first example, it seems you have one factor ('cond') with 3 levels ('cond1', 'cond2', 'cond3') and your formula specifies 'y = cond1 + cond2 + cond3 + ar_noise', how would you handle a case with several factors? Also, your notation means that the formula treats differently numerical factors and categorical factors, maybe it's not an issue maybe it is. Yet another question is how to declare parametric modulations in the parlance of SPM in the formula. 

In R, formulas list factors (i.e. y ~ cond), and their corresponding levels are discovered by the software which then generates an appropriate design matrix. Note that there are several ways to encode categorical variables and ideally, the user should be able to ask questions without knowing the chosen encoding. Maybe worth taking a look at the R package 'contrasts'.

At any rate, we need API examples to see how to handle common situations such as:

1/ simple factorial plan (ANOVA), with two categorical factors A with 2 levels and B with 3 levels. How to generate the additive model without interactions? how to generate the models with A*B interaction? How to specify the contrasts for the main effects of A and B, the interaction if applicable, the contrasts on A (1 d.f.) restricted to B==B1, or to B (2 d.f.) restricted on A==A1

2/ Same with ANCOVA (1 categorical + 1 continuous predictor)

3/ Parametric modulations: let us say that there are two types of stimuli in a experiment (1 factor stim=S1 or S2), and for S1 only, you have 3 continuous variables. How would you write the formula nad the contrasts?

Sorry not to have any proposal at this point.

**Comment by bthirion** [comment]
Hi Christophe

* Regarding the distinction between factors and levels, I agree with you.
* Regarding the manipulation of contrasts and modulation, i would like to rely on existing Python machinery as much as possible. I need to check what's in pandas.
Thx for taking a look.
Best,

Bertrand
